File: dehaze-python/app/service/notification_setting_service.py
from datetime import datetime, time
from typing import Any

from sqlalchemy.ext.asyncio import AsyncSession

from app.models.entity.sys_notification_setting import SysNotificationSetting
from app.repository.notification_setting_repository import notification_setting_repository

DEFAULT_PREFERENCES = {
    "typeChannels": {
        "announcement": {"push": True},
        "business": {"push": False},
        "member": {"push": True},
    },
    "moduleSwitches": {
        "prediction": True,
        "feedback": True,
        "announcement": True,
    },
}


def _format_time(t: time | None) -> str | None:
    if t is None:
        return None
    return t.strftime("%H:%M:%S")


def _to_vo(setting: SysNotificationSetting) -> dict[str, Any]:
    return {
        "pushEnabled": bool(setting.push_enabled),
        "dndEnabled": bool(setting.dnd_enabled),
        "dndStart": _format_time(setting.dnd_start),
        "dndEnd": _format_time(setting.dnd_end),
        "preferences": setting.preferences or DEFAULT_PREFERENCES,
    }


def _deep_merge_preferences(old: dict | None, new: dict | None) -> dict[str, Any]:
    result = dict(old or DEFAULT_PREFERENCES)
    if not new:
        return result
    for key, value in new.items():
        if isinstance(value, dict) and isinstance(result.get(key), dict):
            merged = dict(result[key])
            merged.update(value)
            result[key] = merged
        else:
            result[key] = value
    return result


class NotificationSettingService:
    async def get_or_init(self, db: AsyncSession, user_id: int) -> dict[str, Any]:
        setting = await notification_setting_repository.get_by_user_id(db, user_id)
        if not setting:
            setting = await notification_setting_repository.upsert_by_user_id(db, user_id)
        return _to_vo(setting)

    async def update(self, db: AsyncSession, user_id: int, data: dict[str, Any]) -> None:
        setting = await notification_setting_repository.get_by_user_id(db, user_id)
        if not setting:
            setting = await notification_setting_repository.upsert_by_user_id(db, user_id)

        if "pushEnabled" in data:
            setting.push_enabled = 1 if data["pushEnabled"] else 0
        if "dndEnabled" in data:
            setting.dnd_enabled = 1 if data["dndEnabled"] else 0
        if "dndStart" in data and data["dndStart"]:
            setting.dnd_start = datetime.strptime(data["dndStart"], "%H:%M:%S").time()
        if "dndEnd" in data and data["dndEnd"]:
            setting.dnd_end = datetime.strptime(data["dndEnd"], "%H:%M:%S").time()
        if "preferences" in data and data["preferences"]:
            setting.preferences = _deep_merge_preferences(setting.preferences, data["preferences"])
        await db.flush()


# 单例
notification_setting_service = NotificationSettingService()

File: dehaze-go/internal/service/message/notification_setting_service.go
package message

import (
	"context"
	"encoding/json"
	"time"

	"github.com/earthyzinc/dehaze-go/internal/model"
	"github.com/earthyzinc/dehaze-go/internal/model/bo"
	"github.com/earthyzinc/dehaze-go/internal/model/vo"
	msgrepo "github.com/earthyzinc/dehaze-go/internal/repository/message"
	"github.com/earthyzinc/dehaze-go/pkg/common"
)

const dndTimeFormat = "15:04:05"

type NotificationSettingService struct {
	settingRepo msgrepo.INotificationSettingRepository
}

func NewNotificationSettingService(settingRepo msgrepo.INotificationSettingRepository) *NotificationSettingService {
	return &NotificationSettingService{settingRepo: settingRepo}
}

func (s *NotificationSettingService) Get(ctx context.Context, userID int64) (*vo.NotificationSettingsVO, error) {
	setting, err := s.getOrCreateDefault(ctx, userID)
	if err != nil {
		return nil, err
	}
	return s.toVO(setting), nil
}

func (s *NotificationSettingService) Update(ctx context.Context, userID int64, form *bo.NotificationSettingForm) error {
	setting, err := s.getOrCreateDefault(ctx, userID)
	if err != nil {
		return err
	}

	if form.PushEnabled != nil {
		setting.PushEnabled = boolToInt8(*form.PushEnabled)
	}
	if form.DndEnabled != nil {
		setting.DndEnabled = boolToInt8(*form.DndEnabled)
	}
	if form.DndStart != nil {
		if _, err := time.ParseInLocation(dndTimeFormat, *form.DndStart, time.Local); err != nil {
			return common.NewBizError(common.PARAM_ERROR, "免打扰开始时间格式不正确")
		}
		setting.DndStart = *form.DndStart
	}
	if form.DndEnd != nil {
		if _, err := time.ParseInLocation(dndTimeFormat, *form.DndEnd, time.Local); err != nil {
			return common.NewBizError(common.PARAM_ERROR, "免打扰结束时间格式不正确")
		}
		setting.DndEnd = *form.DndEnd
	}
	if form.Preferences != nil {
		var existing map[string]interface{}
		if setting.Preferences != "" {
			_ = json.Unmarshal([]byte(setting.Preferences), &existing)
		}
		if existing == nil {
			existing = make(map[string]interface{})
		}
		formPrefsBytes, _ := json.Marshal(form.Preferences)
		var formPrefsMap map[string]interface{}
		_ = json.Unmarshal(formPrefsBytes, &formPrefsMap)
		for k, v := range formPrefsMap {
			if newMap, ok := v.(map[string]interface{}); ok {
				if oldMap, ok := existing[k].(map[string]interface{}); ok {
					for k2, v2 := range newMap {
						oldMap[k2] = v2
					}
					continue
				}
			}
			existing[k] = v
		}
		setting.Preferences = toJSONString(existing)
	}

	if err := s.settingRepo.Update(ctx, setting); err != nil {
		return common.WrapBizError(common.DATABASE_ERROR, "更新通知设置失败", err)
	}
	return nil
}

func (s *NotificationSettingService) getOrCreateDefault(ctx context.Context, userID int64) (*model.SysNotificationSetting, error) {
	setting, err := s.settingRepo.FindByUserID(ctx, userID)
	if err != nil {
		return nil, common.WrapBizError(common.DATABASE_ERROR, "查询通知设置失败", err)
	}
	if setting != nil {
		return setting, nil
	}

	defaultPrefs := vo.NotificationPreferences{
		TypeChannels: map[string]vo.TypeChannel{
			"announcement": {Push: true},
			"business":     {Push: false},
			"member":       {Push: true},
		},
		ModuleSwitches: map[string]bool{
			"prediction":   true,
			"feedback":     true,
			"announcement": true,
		},
	}
	setting = &model.SysNotificationSetting{
		UserID:      userID,
		PushEnabled: 1,
		DndEnabled:  0,
		DndStart:    "22:00:00",
		DndEnd:      "08:00:00",
		Preferences: toJSONString(defaultPrefs),
	}

	if err := s.settingRepo.Create(ctx, setting); err != nil {
		return nil, common.WrapBizError(common.DATABASE_ERROR, "创建默认通知设置失败", err)
	}
	return setting, nil
}

func (s *NotificationSettingService) toVO(setting *model.SysNotificationSetting) *vo.NotificationSettingsVO {
	result := &vo.NotificationSettingsVO{
		PushEnabled: setting.PushEnabled == 1,
		DndEnabled:  setting.DndEnabled == 1,
		DndStart:    setting.DndStart,
		DndEnd:      setting.DndEnd,
		Preferences: vo.NotificationPreferences{
			TypeChannels:   make(map[string]vo.TypeChannel),
			ModuleSwitches: make(map[string]bool),
		},
	}
	if setting.Preferences != "" {
		_ = json.Unmarshal([]byte(setting.Preferences), &result.Preferences)
	}
	if result.Preferences.TypeChannels == nil {
		result.Preferences.TypeChannels = make(map[string]vo.TypeChannel)
	}
	if result.Preferences.ModuleSwitches == nil {
		result.Preferences.ModuleSwitches = make(map[string]bool)
	}
	return result
}

func boolToInt8(b bool) int8 {
	if b {
		return 1
	}
	return 0
}

var _ INotificationSettingService = (*NotificationSettingService)(nil)
